File: state_linker.py
from visitor import Visitor
from constructs import INSTATE_SEQ
from compiler_exceptions import CompilerException
from lexer import Lexer, TokenType

class StateReferenceException(CompilerException):
    pass

class StateLinker(Visitor):
    
    def __init__(self):
        self.visiting_statechart = None
        self.visiting_node = None
    
    def visit_ClassDiagram(self, class_diagram): 
        for c in class_diagram.classes :
            c.accept(self)

    def visit_Class(self, c):
        c.statechart.accept(self)
        
    def visit_StateChart(self, statechart):
        self.visiting_statechart = statechart
        for node in statechart.basics + statechart.composites:
            node.accept(self)
                     
    def visit_StateChartNode(self, node):
        self.visiting_node = node
        node.entry_action.accept(self)
        node.exit_action.accept(self)
        for transition in node.transitions :
            transition.accept(self)
            
    def visit_StateChartTransition(self, transition):
        try :
            transition.target.accept(self)
        except StateReferenceException as exception :
            raise StateReferenceException("Transition from <" + self.visiting_node.getFullID() + "> has invalid target. " + exception.message)
        try :
            transition.action.accept(self)
        except StateReferenceException as exception :
            raise StateReferenceException("Transition from <" + self.visiting_node.getFullID() + "> has invalid action. " + exception.message)
        try :
            if transition.guard :
                transition.guard.accept(self)
        except StateReferenceException as exception :
            raise StateReferenceException("Transition from <" + self.visiting_node.getFullID() + "> has invalid guard. " + exception.message)
        
    def visit_StateReference(self, state_path):
        current_node = None #Will be used to find the target state(s)
        split_stack = [] #used for branching

        lx = Lexer()
        lx.input(state_path.path_string)
        token = None
        last_token = None

        while True :
            last_token = token
            token = lx.nextToken() #get next token
            if token == None : #if that next token is None, then we ran out of tokens and break
                break
            
            if current_node == None : #current_node is not set yet or has been reset, the CHILD token can now have a special meaning
                if token.type == TokenType.SLASH :
                    #Root detected
                    current_node = self.visiting_statechart.root
                    #Token consumed so continue
                    continue
                else :
                    current_node = self.visiting_node
                    
            
            if token.type == TokenType.PARENT :
                current_node = current_node.getParentNode()
                if current_node == None :
                    raise StateReferenceException("Illegal use of parent operator at position " + str(token.pos) + " in state reference. Root of statechart reached.")
            elif token.type == TokenType.CURRENT :
                continue
            elif token.type == TokenType.SLASH :
                pass
            elif token.type == TokenType.WORD :
                #try to advance to next child state
                cname = token.val
                found = False
                for child in current_node.children :
                    if child.getName() == cname : 
                        found = True
                        current_node = child
                        break
                if not found :
                    raise StateReferenceException("Refering to non exiting node at posisition " + str(token.pos) + " in state reference.")
            elif token.type == TokenType.LB :
                split_stack.append(current_node)
            elif token.type == TokenType.RB :
                split_stack.pop()
            elif token.type == TokenType.COMMA :
                state_path.target_states.append(current_node)
                if len(split_stack) > 0 :
                    current_node = split_stack[-1]
                else :
                    current_node = None
            
            else :
                raise StateReferenceException("Invalid token at position " + str(token.pos) + ".")

            
        #TODO : better validation of the target! When is it a valid state configuration?
        
        if (len(split_stack) != 0) or (current_node is None) : #RB missing or extra COMMA
            raise StateReferenceException("Missing token at position " + str(last_token.pos) + ".")
        
        state_path.target_states.append(current_node)
            
        if len(state_path.target_states) == 0 :
            raise StateReferenceException("Meaningless state reference.")
        

        

        
    #edit this class out
    def visit_EnterExitAction(self, action):
        if action.action :
            action.action.accept(self)
            
    def visit_Action(self, action):
        for subaction in action.sub_actions :
            subaction.accept(self)
            
    def visit_InStateCall(self, call):
        try :
            call.target.accept(self)
        except StateReferenceException :
            raise StateReferenceException("Invalid state reference for " + INSTATE_SEQ + " call.")
        
    def visit_Assign(self, assign):
        assign.expression.accept(self)

File: csharp_sccd_compiler/Visitors/StateLinker.cs
using System;
using System.Linq;
using System.Collections.Generic;

namespace csharp_sccd_compiler
{
    public class StateLinker : Visitor
    {
        StateChart visiting_statechart;
        StateChartNode visiting_node;
        Lexer lexer;

        public StateLinker()
        {
            this.lexer = new Lexer();
        }

        public override void visit(ClassDiagram class_diagram)
        {
            foreach (Class c in class_diagram.classes)
                c.accept(this);
        }

        public override void visit(Class c)
        {
            c.statechart.accept(this);
        }

        public override void visit(StateChart statechart)
        {
            this.visiting_statechart = statechart;
            foreach (StateChartNode node in statechart.basics.Concat(statechart.composites))
                node.accept(this);
        }
        
        public override void visit(StateChartNode node)
        {
            this.visiting_node = node;
            node.enter_action.accept(this);
            node.exit_action.accept(this);
            foreach (StateChartTransition transition in node.transitions)
                transition.accept(this);
        }
               

        public override void visit(StateChartTransition transition)
        {
            try
            {
                transition.target.accept(this);
            }
            catch (StateReferenceException exception)
            {
                throw new StateReferenceException(string.Format("Transition from <{0}> has invalid target.", this.visiting_node.full_name), exception);
            }
            try
            {
                transition.action.accept(this);
            }
            catch (StateReferenceException exception)
            {
                throw new StateReferenceException(string.Format("Transition from <{0}> has invalid action.", this.visiting_node.full_name), exception);
            }
            try
            {
                if (transition.guard != null)
                    transition.guard.accept(this);
            }
            catch (StateReferenceException exception)
            {
                throw new StateReferenceException(string.Format("Transition from <{0}> has invalid guard.", this.visiting_node.full_name), exception);
            }
        }

        public override void visit(StateReference state_reference)
        {
            state_reference.target_nodes = new List<StateChartNode>();
            StateChartNode current_node = null; //Will be used to find the target state(s)
            Stack<StateChartNode> split_stack = new Stack<StateChartNode>(); // user for branching
            this.lexer.setInput(state_reference.state_reference_string);

            foreach (Token token in this.lexer.iterateTokens())
            {
                if (current_node == null) //current_node is not set yet or has been reset, the CHILD token can now have a special meaning
                {
                    if (token.type == Token.Type.SLASH) 
                    {
                        current_node = this.visiting_statechart.root; //Root detected
                        continue;
                    }
                    else
                        current_node = this.visiting_node;
                }

                if (token.type == Token.Type.DOT)
                {
                    Token next_token = this.lexer.nextToken();//Advance to next token
                    if (next_token == null || next_token.type == Token.Type.SLASH)
                        continue;//CURRENT operator "." detected
                    else if (next_token.type == Token.Type.DOT)
                    {
                        next_token = this.lexer.nextToken();//Advance to next token
                        if (next_token == null || next_token.type == Token.Type.SLASH)
                        {
                            current_node = current_node.parent; //PARENT operator ".." detected
                            if (current_node == null)
                                throw new StateReferenceException(string.Format("Illegal use of PARENT \"..\" operator at position {0} in state reference. Root of statechart reached.", token.pos));
                        }
                        else
                            throw new StateReferenceException(string.Format("Illegal use of PARENT \"..\" operator at position {0} in state reference.", token.pos));
                    }
                    else
                        throw new StateReferenceException(string.Format("Illegal use of CURRENT \"..\" operator at position {0} in state reference.", token.pos));
                }
                else if (token.type == Token.Type.SLASH)
                    continue;
                else if (token.type == Token.Type.WORD)
                {
                    //Trying to advance to next child state
                    string child_name = token.val;
                    bool found = false;
                    foreach (StateChartNode child in current_node.children)
                    {
                        if (child.name == child_name)
                        {
                            found = true;
                            current_node = child;
                            break;
                        }
                    }
                    if (!found)
                        throw new StateReferenceException(string.Format("Refering to non exiting node at posisition {0} in state reference.", token.pos));
                }
                else if (token.type == Token.Type.LBRACKET)
                    split_stack.Push(current_node);
                else if (token.type == Token.Type.RBRACKET)
                {
                    if (split_stack.Count > 0)
                        split_stack.Pop();
                    else
                        throw new StateReferenceException(string.Format("Invalid token at position {0} in state reference.", token.pos));
                }
                else if (token.type == Token.Type.COMMA)
                {
                    state_reference.target_nodes.Add(current_node);
                    if (split_stack.Count > 0)
                        current_node = split_stack.Peek();
                    else
                        current_node = null;
                }
                else
                    throw new StateReferenceException(string.Format("Invalid token at position {0} in state reference.", token.pos));
            }

            if (split_stack.Count != 0 || current_node == null) //RB missing or extra COMMA
                throw new StateReferenceException("State reference ends unexpectedly.");

            //TODO better validation of the target! When is it a valid state configuration?
            foreach (StateChartNode node in state_reference.target_nodes)
            {
                if (object.ReferenceEquals(current_node,node))
                    throw new StateReferenceException("A state reference can't reference the same node more than once.");
                if (node.isDescendantOrAncestorOf(current_node))
                    throw new StateReferenceException("A state reference can't reference a node and one of its descendants.");
            }

            state_reference.target_nodes.Add(current_node);

            if (state_reference.target_nodes.Count == 0)
                throw new StateReferenceException("Meaningless state reference.");
        }

        public override void visit(EnterExitAction enter_exit_action)
        {
            if (enter_exit_action.action != null)
                enter_exit_action.action.accept(this);
        }

        public override void visit(InStateCall in_state_call)
        {
            try
            {
                in_state_call.state_reference.accept(this);
            }
            catch (StateReferenceException exception)
            {
                throw new StateReferenceException(string.Format("Invalid state reference for {0} macro.", Constants.INSTATE_SEQ), exception);
            }
        }

        public override void visit(Assign assign)
        {
            assign.expression.accept(this);
        }
    }
}

